cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -126,6 +131,84 @@ public CloseableIterator<StructuredRow> scan(Range keyRange, int limit) throws I
                              limit);
   }
 
+  @Override
+  public CloseableIterator<StructuredRow> scan(Range keyRange, int limit, Field<?> orderByField, SortOrder sortOrder)
+    throws InvalidFieldException, IOException {
+    LOG.trace("Table {}: Scan range {} with limit {}, sort field {} and sort order {}", schema.getTableId(),
+              keyRange, limit, orderByField, sortOrder);
+    fieldValidator.validateField(orderByField);
+    if (!schema.isIndexColumn(orderByField.getName())) {
+      throw new InvalidFieldException(schema.getTableId(), orderByField.getName(), "is not an indexed column");
+    }
+    ScannerIterator scannerIterator = new ScannerIterator(getScanner(keyRange), schema);
+    LimitIterator limitIterator = new LimitIterator(Collections.singleton(scannerIterator).iterator(),
+                                                    limit);
+    PriorityQueue<StructuredRow> sorted = new PriorityQueue<>(sortOrder.equals(SortOrder.ASC) ?
+                                                                (row1, row2) ->
+                                                                  compareRowFields(row1, row2, orderByField)
+                                                                : (row1, row2) ->
+      compareRowFields(row2, row1, orderByField));
+
+    while (limitIterator.hasNext()) {
+      StructuredRow row = limitIterator.next();
+      sorted.offer(row);
+    }
+    // return an iterator for the sorted elements
+    return new AbstractCloseableIterator<StructuredRow>() {
```

Comment:
PriorityQueue is already Iterable, we should be delegating instead of implementing logic. Can create a class that looks something like:

```
public class DelegatingCloseableIterator<T> implements Iterator<T> {
  private final Iterator<T> delegate;

  public DelegatingCloseableIterator(Iterator<T> delegate) {
    this.delegate = delegate;
  }

  @Override
  public boolean hasNext() {
    return delegate.hasNext();
  }

  ...

  @Override
  public void close() {
    // no-op
  }
}
```

Reply:
Shouldn't it be like this:

```/**
   * Returns elements from an iterator
   */
  @VisibleForTesting
  static final class DelegatingCloseableIterator extends AbstractCloseableIterator<StructuredRow> {
    private final Iterator<StructuredRow> delegate;

    DelegatingCloseableIterator(Iterator<StructuredRow> delegate) {
      this.delegate = delegate;
    }

    @Override
    protected StructuredRow computeNext() {
      while (delegate.hasNext()) {
        return delegate.next();
      }
      return endOfData();
    }

    @Override
    public void close() {
      delegate.remove();
    }
  }```

Since we are iterating over StructuredRow and intend to  extend `AbstractCloseableIterator`

Reply:
the comment is that it shouldn't be extending AbstractCloseableIterator.